src/database-postgresql.cpp

```diff
@@ -0,0 +1,238 @@
+	}
+
+	infostream << "PostgreSQL Database: Version " << m_pgversion << " Connection made." << std::endl;
+}
+
+void Database_PostgreSQL::verifyDatabase()
+{
+	if (PQstatus(m_conn) == CONNECTION_OK)
+		return;
+
+	PQreset(m_conn);
+	ping();
+}
+
+void Database_PostgreSQL::ping()
+{
+	if (PQping(m_connect_string.c_str()) != PQPING_OK) {
+		throw DatabaseException(std::string(
+			"PostgreSQL database error: ") +
+			PQerrorMessage(m_conn));
+	}
+}
+
+bool Database_PostgreSQL::initialized() const
+{
+	return (PQstatus(m_conn) == CONNECTION_OK);
+}
+
+void Database_PostgreSQL::initStatements()
+{
+	prepareStatement("read_block",
+			"SELECT block_data FROM blocks "
+					"WHERE posX = $1::int AND posY = $2::int AND posZ = $3::int");
+
+	prepareStatement("write_block",
+			"INSERT INTO blocks (posX, posY, posZ, block_data) VALUES "
+					"($1::int, $2::int, $3::int, $4::bytea) "
+					"ON CONFLICT ON CONSTRAINT blocks_pkey DO UPDATE SET block_data = $4::bytea");
+
+	prepareStatement("delete_block", "DELETE FROM blocks WHERE posX = $1::int AND posY = $2::int AND posZ = $3::int");
+
+	prepareStatement("list_all_loadable_blocks", "SELECT posX, posY, posZ FROM blocks");
+}
+
+void Database_PostgreSQL::prepareStatement(const std::string &name, const std::string &sql)
+{
+	checkResults(PQprepare(m_conn, name.c_str(), sql.c_str(), 0, NULL));
+}
+
+PGresult *Database_PostgreSQL::checkResults(PGresult *result, bool clear /*= true*/)
+{
+	ExecStatusType statusType = PQresultStatus(result);
+
+	if ((statusType != PGRES_COMMAND_OK && statusType != PGRES_TUPLES_OK) ||
+		statusType == PGRES_FATAL_ERROR) {
+		throw DatabaseException(std::string(
+			"PostgreSQL database error: ") +
+			PQresultErrorMessage(result));
+	}
+
+	if (clear)
+		PQclear(result);
+
+	return result;
+}
+
+void Database_PostgreSQL::createDatabase()
+{
+	PGresult *result = checkResults(PQexec(m_conn,
+		"SELECT relname FROM pg_class WHERE relname='blocks';"),
+		false);
+
+	// If table doesn't exist, create it
+	if (!PQntuples(result)) {
+		static const char* dbcreate_sql = "CREATE TABLE blocks ("
+			"posX INT NOT NULL,"
+			"posY INT NOT NULL,"
+			"posZ INT NOT NULL,"
+			"block_data BYTEA,"
+			"PRIMARY KEY (posX,posY,posZ)"
+		");";
+		checkResults(PQexec(m_conn, dbcreate_sql));
+	}
+
+	PQclear(result);
+
+	infostream << "PostgreSQL: Game Database was inited." << std::endl;
+}
+
+
+void Database_PostgreSQL::beginSave()
+{
+	verifyDatabase();
+	checkResults(PQexec(m_conn, "BEGIN;"));
+}
+
+void Database_PostgreSQL::endSave()
+{
+	checkResults(PQexec(m_conn, "COMMIT;"));
+}
+
+bool Database_PostgreSQL::saveBlock(const v3s16 &pos, const std::string &data)
+{
+	verifyDatabase();
+
+	// This case is manually handled because we manipulate blob datas
+	std::string x = itos(pos.X), y = itos(pos.Y), z = itos(pos.Z);
+	const char *values[] = {x.c_str(), y.c_str(), z.c_str(), data.c_str()};
+	const int valLen[] = {sizeof(int), sizeof(int), sizeof(int), (int)data.size()};
```

Comment:
What is the size of an int in postgresql?
Do you realize the size of an int varies based on platforms?  That means map dbs shared between different computers might not work due to a size mismatch.